core/src/test/java/greencity/controller/TipsAndTricksCommentControllerTest.java

```diff
@@ -0,0 +1,182 @@
+import greencity.entity.User;
+import greencity.service.TipsAndTricksCommentService;
+import greencity.service.UserService;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.mockito.junit.jupiter.MockitoSettings;
+import org.mockito.quality.Strictness;
+import org.springframework.context.annotation.Import;
+import org.springframework.data.domain.PageRequest;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.web.PageableHandlerMethodArgumentResolver;
+import org.springframework.http.MediaType;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.test.web.servlet.setup.MockMvcBuilders;
+
+import java.security.Principal;
+
+import static org.mockito.Mockito.*;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+@ExtendWith(MockitoExtension.class)
+@MockitoSettings(strictness = Strictness.LENIENT)
+@ContextConfiguration
+@Import(SecurityConfig.class)
+public class TipsAndTricksCommentControllerTest {
+    private static final String tipsAndTricksCommentLink = "/tipsandtricks/comments";
+    private MockMvc mockMvc;
+    @InjectMocks
+    private TipsAndTricksCommentController tipsAndTricksCommentController;
+    @Mock
+    private TipsAndTricksCommentService tipsAndTricksCommentService;
+    @Mock
+    private UserService userService;
+
+    @BeforeEach
+    void setup() {
+        this.mockMvc = MockMvcBuilders.standaloneSetup(tipsAndTricksCommentController)
+                .setCustomArgumentResolvers(new PageableHandlerMethodArgumentResolver())
+                .build();
+    }
+
+    @Test
+    public void saveTest() throws Exception {
+        Principal principal = Mockito.mock(Principal.class);
+        when(principal.getName()).thenReturn("Liam.Johnson@gmail.com");
+        when(userService.findByEmail(anyString())).thenReturn(new User());
+
+        String content = "{\n" +
+                "  \"parentCommentId\": 0,\n" +
+                "  \"text\": \"string\"\n" +
+                "}";
+
+        mockMvc.perform(post(tipsAndTricksCommentLink + "/{tipsAndTricksId}", 1)
+                .principal(principal)
+                .contentType(MediaType.APPLICATION_JSON)
+                .content(content))
+                .andExpect(status().isCreated());
+
+        ObjectMapper mapper = new ObjectMapper();
+        AddTipsAndTricksCommentDtoRequest addTipsAndTricksCommentDtoRequest = mapper.readValue(content, AddTipsAndTricksCommentDtoRequest.class);
+
+        verify(userService).findByEmail(eq("Liam.Johnson@gmail.com"));
+        verify(tipsAndTricksCommentService)
```

Comment:
`eq(new User())` is bad. In fact it should be equal not to a new user, but to the user returned by `userService.findByEmail`. Looks like the test pass because `equals` is not implemented.

Reply:
User class has such annotation:
@EqualsAndHashCode(
    exclude = {"places", "comments", "verifyEmail", "addedPlaces", "favoritePlaces", "ownSecurity", "refreshTokenKey",
        "verifyEmail", "estimates", "restorePasswordEmail", "addedEcoNews", "addedTipsAndTricks"})

Replaced new User() to User from ModelUtils

Reply:
Yes, much better now. Looks like before the check `new User().equals(new User())` was returning true, which is not really true.